### ISSUE TITLE
chunkhash does not change when module ids do

### ISSUE THREAD

**Issue by virajsanghvi** [webpack/webpack#1856]
I have been unable to create a simple repro of this issue, but I'm hoping by describing it, people may have an idea of what's going on. This sounds similar to #1155.

We have two entries, one that just requires some images using url-loader (entry A), and another includes our javascript app (entry B).

After making some changes that should only effect entry B, we built everything, and found that entry A had the same file name (same chunkhash) as before, even though module ids had changed. For example, before:
```javascript
!function(e){function c(r){if(n[r])return n[r].exports;var a=n[r]={exports:{},id:r,loaded:!1};return e[r].call(a.exports,a,a.exports,c),a.loaded=!0,a.exports}var n={};return c.m=e,c.c=n,c.p="",c(0)}({0:function(e,c,n){"use strict";n(652),n(650),n(649)},649:function(e,c,n){e.exports=n.p+"images/favicon.d6a3e4a5239105abdca2874ee0b29e87.png"},650:function(e,c,n){e.exports=n.p+"images/icons/image1.94cd651e18cc108dbcac932ca11f87e0.svg"},652:function(e,c,n){e.exports=n.p+"images/image2.3d0c754f88b3aa29f88f1422f158649d.png"}});
```

after:
```javascript
!function(e){function c(r){if(n[r])return n[r].exports;var a=n[r]={exports:{},id:r,loaded:!1};return e[r].call(a.exports,a,a.exports,c),a.loaded=!0,a.exports}var n={};return c.m=e,c.c=n,c.p="",c(0)}({0:function(e,c,n){"use strict";n(653),n(651),n(650)},650:function(e,c,n){e.exports=n.p+"images/favicon.d6a3e4a5239105abdca2874ee0b29e87.png"},651:function(e,c,n){e.exports=n.p+"images/icons/image1.94cd651e18cc108dbcac932ca11f87e0.svg"},653:function(e,c,n){e.exports=n.p+"images/image2.3d0c754f88b3aa29f88f1422f158649d.png"}});
```

Just fyi, plugins we're using:
```
CommonsChunkPlugin
UglifyJsPlugin
DedupePlugin
OccurenceOrderPlugin(true)
```

Assuming this (https://webpack.github.io/docs/plugins.html) specifies the order of steps, hashing occurs between ```before-hash``` and ```after-hash``` while I'm assuming module ids are last changed at ```after-optimize-module-ids``` (which occurs before the hash). UglifyJsPlugin is happening at ```optimize-chunk-assets```, although as this happens after the chunkhash is calculated, it shouldn't matter here.

My guess is that chunkhash is calculated solely on content, but this is an issue if its intended to be used for long term caching. If anyone can point to where this is happening, I don't mind taking a look.

I'll continue to try to find a simple repro. I should note that this isn't probably isn't specific to my setup above, and just presented itself that way.

**Comment by sokra** [comment]
https://github.com/webpack/webpack/blob/master/lib/Compilation.js#L530-L550

The hash is created after the ids are calculated, and hash includes module ids...
So this should not happen... Are you using the current version?

**Comment by virajsanghvi** [comment]
I'm using 1.12.9, and the diff between that and 1.12.10 doesn't appear like it would resolve this, but I can try it out.

This does seem peculiar, and likely something in my configuration. Do you think any of the plugins might potentially have a side effect that would present itself like this? Most likely I'll try to disable things in my complicated setup to narrow down what might be the root cause, so feel free to close, and I can reopen if I find something.

**Comment by riston** [comment]
I am having similar problem, seems like the output for chunk is not used after all, when filename property exists ? 

`
  output: {
  	path: "./build",
    filename: "[name].[hash].js",
    chunkFilename: "[name][chunkhash].js",
  },
`
I made simple example to show the problem: https://github.com/riston/webpack-chunkhash-problem

**Comment by L3au** [comment]
I also have this problem and I use `webpack-md5-hash` plugin, but it still may occur.

**Comment by stocks29** [comment]
We were having the same problem. We were able to work around it by including the `NamedModulesPlugin` in our config.

**Comment by alexindigo** [comment]
Seeing same issue. Different ids, everything else is the same including `chunkhash`.

![same hashes](http://ia.gs/0O0X2M0Q1h1J/Image%202016-11-08%20at%202.58.41%20PM.png)

Using `[chunkhash]` for the filename and `new WebpackMd5Hash()` in the plugins section.

**Comment by alexindigo** [comment]
On more investigation, seems like `WebpackMd5Hash` plugin (with node@4) affects ids in the output file.

Ids (same file as on screenshot) generated per node version / md5 plugin:

| node |  w/ md5 | w/o md5 |
| :--     |   :--        | :--            |
| v0.12 | 1101      |  1101         |
| v4.6   | 1103     |  1101         |

**Comment by alexindigo** [comment]
Funny, it produced exactly same hash, without pathinfo comments, so it's definitely not the _file's_ content.

```
cat public/js/2.0603ab240ab51473dd48.js 
webpackJsonp([2],{1103:function(e,r,t){var o;o=function(){var e={getOneMonthAgoDate:function(){var e=new Date,r=e.getDate(),t=e.getMonth(),o=e.getFullYear();return 0===t&&(t=12,o--),r<10&&(r="0"+r),t<10&&(t="0"+t),o+"-"+t+"-"+r}};return e}.call(r,t,r,e),!(void 0!==o&&(e.exports=o))},1104:function(e,r,t){var o,a;o=[t(2),t(10),t(1103),t(425),t(424)],a=function(e,r,t,o,a){function p(p,n,d){function i(e){var r;r={modulesSpec:y},n.api(l,r,function(r,t){return r?void e(r):void e(null,t)})}function s(r){return e.mapObject(r,function(e,r){return e.propertyUrl=p.propertyUrl||"",e})}function u(r){return e.extend({},r,{propertyAddress:p.propertyAddress})}function c(e){var r,t;t=u(s(e)),r=new o({app:n,model:new a(t,{app:n}),el:d}),r.render()}var y,l="/api/listings_nearby",m=3,f=1,v="localInfoWeb";return p&&"object"==typeof p?n&&"object"==typeof n?d&&"object"==typeof d?(y={schoolsSummary:{model:"schools_summary",params:{type:p.propertyType,stateCode:p.propertyStateCode,id:p.propertyId}},amenitiesSummary:{model:"amenities_summary",params:{radius:f,limit:m,lat:p.propertyLat,lon:p.propertyLon}},crimesSummary:{model:"crimes_summary",params:{radius:f,lat:p.propertyLat,lon:p.propertyLon,date_start:t.getOneMonthAgoDate()}},demographicsSummary:{model:"demographics_summary",params:{lat:p.propertyLat,lon:p.propertyLon,client:v}}},{show:function(){i(function(e,t){return e?void r.error("Unable to fetch the nearby data: ",e):void c(t)})}}):void r.error("Expected `el` as jQuery object."):void r.error("Expected `app` as a object."):void r.error("Expected `propertyData` as a object.")}return p}.apply(r,o),!(void 0!==a&&(e.exports=a))},1101:function(e,r,t){var o,a;o=[t(1104)],a=function(e){function r(r,t,o){e({propertyType:t.metadata.type,propertyStateCode:t.metadata.stateCode,propertyId:t.id,propertyLat:t.location.latitude,propertyLon:t.location.longitude,propertyUrl:t.pdpURL,propertyAddress:t.address.components.streetAddress},r,o).show()}return r}.apply(r,o),!(void 0!==a&&(e.exports=a))}});
```

**Comment by alexindigo** [comment]
Ok, looks like we're barking at the wrong tree: https://github.com/erm0l0v/webpack-md5-hash/issues/7

**Comment by alexindigo** [comment]
In the same time, `webpack-md5-hash` plugin gets raw modules' content, no ids, no minification.

```
WebpackMd5Hash.prototype.apply = function(compiler) {
    compiler.plugin("compilation", function(compilation) {
        compilation.plugin("chunk-hash", function(chunk, chunkHash) {

            var source = chunk.modules.sort(compareModules).map(getModuleSource).reduce(concatenateSource, ''); // we provide an initialValue in case there is an empty module source. Ref: http://es5.github.io/#x15.4.4.21
            var chunk_hash = md5(source);

            chunkHash.digest = function () {
                return chunk_hash;
            };
        });
    });
};
```

Not sure if it's intended/expect use.

**Comment by alexindigo** [comment]
Ok, seems like it's due to extra (obtrusive) sorting performed by webpack-md5-hash which messes up with `OccurrenceOrderPlugin`. I created alternative version without "manual" sorting – https://www.npmjs.com/package/webpack-chunk-hash (also it uses crypto module, instead of js implementation of md5, so it should speed up your builds as well). It did fix my issue.

**Comment by jeromecovington** [comment]
@alexindigo do you know how I can create a bundle with changed module ids, to verify the work you put into your alternate version?

**Comment by jeromecovington** [comment]
NM, I believe the steps to reproduce in https://github.com/erm0l0v/webpack-md5-hash/issues/7 will be adequate.

**Comment by alexindigo** [comment]
Yeah, for me it was just increased number of files in the main chunk.

**Comment by Poky85** [comment]
I have created simple demo to test different hashing algorithms https://github.com/Poky85/webpack-chunkhash-demo

Given first tests none of tested algorithms gives 100% correct builds in all test cases.

Maybe we can check different Webpack configurations. Feel free to fork my demo.

**Comment by Poky85** [comment]
Summarized test using https://github.com/Poky85/webpack-chunkhash-demo

https://www.icloud.com/numbers/0cEAfcZiNPPwfzGMm4_n160NQ#webpack-chunkhash-demo-resuts

Would be nice to make test automatized :-)

**Comment by sokra** [comment]
Could you test it without `ChunkManifestPlugin`?

**Comment by sokra** [comment]
It's would be great to automate this test. Best as webpack test case.

There are already tests which check the hashes on update. i. e. this one: https://github.com/webpack/webpack/blob/master/test/hotCases/hash/separate-css-file/index.js
but with HMR enabled it behave a bit different (main chunk always change).

Maybe you could do a PR?

**Comment by alexindigo** [comment]
@Poky85 very cool test and spreadsheet. Thank you. I'll borrow your trst cases for chunk-hash plugin. Thanks.

**Comment by alexindigo** [comment]
Interesting what this kind of change means in the real life:
![master->testcase1](http://ia.gs/2l3k0b0V3g2J/Image%202016-11-12%20at%208.54.52%20PM.png)

**Comment by alexindigo** [comment]
Looking more into `webpack-chunk-hash` example of your demo, it's interesting, that source code for `entry-2` hasn't changed between `master` and `testcase1` (failed example), it did change in the compiled version though. 

And plugins listening on `chunk-hash` event, receive source code rather than compiled version.

@sokra is there a way to get compiled version for `chunk-hash`? Thanks.

**Comment by alexindigo** [comment]
Ok, I think I've found the workaround. Not sure how generic it is, but it does solve this example:

master -> testcase1:

![master -> testcase1](http://ia.gs/1Q0z3v0o133I/Image%202016-11-12%20at%2010.27.48%20PM.png)

testcase1 -> testcase2:

![testcase1 -> testcase2](http://ia.gs/1t2r1v180q2U/Image%202016-11-12%20at%2010.29.32%20PM.png)

**Comment by alexindigo** [comment]
@Poky85 I published `webpack-chunk-hash@0.3.0` that should pass your demo. Can you take a look? Thanks.

PS. Now it checks for module.dependencies' ids and their order.

**Comment by Poky85** [comment]
@alexindigo Now `webpack-chunk-hash@0.3.0` pass all tests in my demo.

Definitely hashing plugin should consider compiled version of chunk for hashing. Because in case you have 2 or more chunks that contain same modules then module IDs can change on every build resulting in different chunk even if source code didn't change.

I will test your plugin in real world application and will thinking about more test cases (for example I have experienced that some node_module(s) breaks hashing etc.)

I updated my demo https://github.com/Poky85/webpack-chunkhash-demo so we can easily test different Webpack configurations.

**Comment by alexindigo** [comment]
@Poky85 Cool. Thanks for all your work.

Meanwhile, I'm working on integrating your tests into automatic tests in webpack-chunk-hash.

**Comment by Poky85** [comment]
@sokra Yes, default Webpack chunkhash algorithm pass all tests in https://github.com/Poky85/webpack-chunkhash-demo when `ChunkManifestPlugin` is not used.

Interesting that chunkhash for given chunk is same – it doesn't matter whether `ChunkManifestPlugin` is applied or not. Although output chunk with or without plugin differs so chunkhash should differ too.

**Comment by alexindigo** [comment]
@Poky85 btw, I think it'd be better to mark `no change` -> `change` fail with non-red color, since it doesn't break the app, only degrades the performance, comparing to `change` -> `no change` that actually break the app. Thanks.

**Comment by Poky85** [comment]
> I think it'd be better to mark no change -> change fail with non-red color, since it doesn't break the app, only degrades the performance

@alexindigo It depends. From one point of view it's fail because chunkhash plugin goal is to create unique hash to make long-term cache working. So broken caching makes plugin useless in my opinion.

**Comment by alexindigo** [comment]
@Poky85 for sure, I'm not calling it a Pass, it is a Fail. Although I this kind of Fail as not as bad, when app throws error in the browser and stops functioning. The latter is my experience with `change` -> `no change` big.

**Comment by Poky85** [comment]
@alexindigo Btw in my app I have interesting problem. Sometimes same chunk differs between 2 builds although source code is same. I found that Webpack changes module IDs between those 2 builds (in my case all IDs are increased by 1).

This is screenshot of module source used for hashing in your `webpack-chunk-hash` plugin (with some extra console.log :-))

![image](https://cloud.githubusercontent.com/assets/1475938/20271601/9ab139ca-aa8b-11e6-9b99-bb02a86d3256.png)

Pretty sure I have seen this is issue somewhere...

**Comment by alexindigo** [comment]
This exactly what brought me in tis whole webpack hashing thingy. Seems like it's doing of OrderOccurence plugin, and also md5-hash us mutating modules array, which added more noise.

But despite those things, it's hard problem in general, if you replace named references with numbers, then you have modules like 1, 2, 3, 4, 5 and when you add one more module it will affect some of the modules (unless you keep some kind of state between builds). 

And when I played with NamedChunk plugin, I found that it's just creating extra mapping, but ids still exist internally, so the problem doesn't go away completely. But this is webpack's design.

@sokra is there a way to get rid of module ids and just use their names/paths (similar to what requirejs does)?

**Comment by Kovensky** [comment]
Save the used module IDs with the [`recordsPath` option](https://webpack.github.io/docs/configuration.html#recordspath-recordsinputpath-recordsoutputpath). This makes webpack always pick the same ID for the same module, as recorded in the records file.

**Comment by Poky85** [comment]
@Kovensky Perfect! When using [`recordsPath` option](https://webpack.github.io/docs/configuration.html#recordspath-recordsinputpath-recordsoutputpath) Webpack keeps module IDs persistent across builds.

But seems that `webpack.optimize.DedupePlugin()` breaks module IDs persistence even when using `recordsPath`.

At least in my app I had to disable DedupePlugin (see my config) https://gist.github.com/Poky85/c7f1d2e523b60ff588f44b2d884050a5

cc @alexindigo

**Comment by alexindigo** [comment]
Thanks. We're yet to adopt recordPath, but thinking about it. What's your take on checking in records file?

And we do use Dedupe, not to soove any specific problem, but as insurance for the future.

@Poky85 I'll borrow few things from your config, thanks. :)

**Comment by nmaves** [comment]
We are super interested in this as well.  I actually opened up an issue with the new webpack 2 docs site to help explain how and when to use the recordsPath.json file.  We really want to know when to commit this file into git.  It changes for us between a DEV and PROD build so we are really confused on when to save the build "state".

**Comment by mnpenner** [comment]
I'm experiencing this issue too in webpack 2.3.2. I can reproduce it 100% of the time by modifying a file that gets put into its own chunk and recompiling. When I do this, the entry chunk should change too because I'm *not* extracting the chunk manifest.

----------------

The problem actually goes away when I *remove* this line:

```js
new WebpackChunkHash({algorithm: 'md5', digest: 'hex'})
```
So perhaps this was fixed in core webpack and `WebpackChunkHash` is now bugged?

**Comment by kingatlas** [comment]
Hello,
I'm having a similar issue related to DLLReferencePlugin.
webpack chunkhash algorithm does not take is consideration ids used by DllReferencePlugin to point to DLL modules located on external bundle.

case 1: app references the jquery library that is bundled in the dll/dist/lib.bundle.js with the id=1 => the generated bundle has the hash 3831aaef113a5b3a187a

case 2: app references the jquery library that is bundled in the dll/dist/lib.bundle.js with the id=4 => the generated bundle has the hash 3831aaef113a5b3a187a

expected result: The two bundles must have different hashes since they have different content.

I have created a sample project to illustrate that issue here: 
https://github.com/kingatlas/webpack-hashing/blob/webpack2/README.md

![image](https://cloud.githubusercontent.com/assets/6798054/24806271/9b21bbc4-1bb4-11e7-93ed-0734fca0a558.png)

**Comment by alexindigo** [comment]
@mnpenner Looks like webpack@2 does better job at hashing now. Have you tried other edge cases from the comments above? To see If webpack2 handles them.

@kingatlas Are you seeing this behavior with webpack-chunk-hash or without? Can you see the difference if you add/remove webpack-chunk-hash?

Thank you.

**Comment by kingatlas** [comment]
@alexindigo There is no difference if I use webpack-chunk-hash (I tried with webpack 2). The hashes are similar for both cases (4f9600f39d39d961f16d).

**Comment by roscoe054** [comment]
@kingatlas The [webpack-hashed-module-id-plugin][1] may solve your problem. It generate module id by its content, so the app.bundle.js content in two cases can be identical.

I've forked your demo and you could try it [here][2].

[1]:https://github.com/roscoe054/webpack-hashed-module-id-plugin
[2]:https://github.com/roscoe054/webpack-hashing

**Comment by kingatlas** [comment]
@roscoe054  thanks for the trial, it is indeed a **good solution for my problem**. Since the module id is not sequential anymore and is independent from the order of declaration.
Do you know if it works for webpack2?

**Comment by roscoe054** [comment]
@kingatlas Actually it's a plugin in [webapck2][1], you can use it directly. 

The independent module is just a modified version which compatible with webpack1. If you use webpack2, you don't need it.

[1]: https://github.com/webpack/webpack/blob/93ac8e9c3699bf704068eaccaceec57b3dd45a14/lib/HashedModuleIdsPlugin.js

**Comment by stevegibbings** [comment]
I've been trying out some of these proposals with results I wasn't expecting but maybe I'm not clear on the use case or intended outcome. I'm after chuckhashes that do not remain the same when a new chunk is introduced which only alters the module ids. For my app and vendor bundle I don't expect them to change as a result of the additional chunk as I use chunk-manifest-webpack-plugin.

With webpack@2.6.1 -

using webpack-hashed-module-id-plugin I get the first few chunk's hashes unchanged but presumably when the new chunk is output the hashes change, even for those existing unchanged chunks that follow. Also the app and vendor hashes change when I wouldn't expect them to. Looking inside my app bundle there is a can't-fine-module promise that has text strings for the original module names that the chunks were generated for. I though extracting the chunk manifest would prevent those kinds of changes?

adding recordsPath option results in the chunks all being seen as new and the [id] in the file name starts from the prior builds id. That means all my chunks have new filenames and all the hashes change.

I can see why my hashes change as the ids even with webpack-hashed-module-id-plugin change and the app bundle has the names of the modules in so they get a new one with my added chunk.

Assuming there are no webpack or plugin bugs here and if (big if) my config is right, should I be expecting  this behaviour?

Thanks

**Comment by roscoe054** [comment]
@stevegibbings It's strange that the existing unchanged chunks are influenced. I can't reproduce on my local project. Can you provide a demo so that I can find out what has happened.

**Comment by stevegibbings** [comment]
Hi. Yes I had started one already. I'll post the GitHub link back soon. Thanks very much.

**Comment by stevegibbings** [comment]
@roscoe054 My initial post wasn't accurate sorry. I've done some more investigation and put together [an example](https://github.com/stevegibbings/webpack-hash-example). I am however still getting hash changes when the file contents are the same meaning I lose the benefit of long term cache.

TL;DR; Using the recommended webpack approach for long term caching I don't get and hash clashes where the hashes didn't change but the file contents did. What I do still get are, IMO, unnecessary hash changes for unaltered output file contents namely vendor & app, vendor being my biggest issue as it's the longest living and biggest bundle. Combinations of adding and removing webpack-chunk-hash-plugin and hashed-module-id-plugin seem to get me closer to an unchanged vendor bundle but not confident those are consistent and stable across untested use cases.

I'm very open to suggestions or issues in my approach or understand being pointed out. I would like to try to get time to look into solutions for these that I can contribute.

Thanks

**Comment by florian-bd** [comment]
Hi, has anyone found a solution other than recordsPath for stable chunk IDs? 
In my project we are adding and removing chunks all the time so it is a problem to have to invalidate users cache everytime this happens.
Also it is very hard to have a proper way of handling this recordsPath file.
